core/src/main/java/io/grpc/internal/StatsTraceContext.java

```diff
@@ -0,0 +1,220 @@
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.census.CensusContext;
+import com.google.census.CensusContextFactory;
+import com.google.census.MetricMap;
+import com.google.census.MetricName;
+import com.google.census.RpcConstants;
+import com.google.census.TagKey;
+import com.google.census.TagValue;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import com.google.common.base.Supplier;
+
+import io.grpc.Metadata;
+import io.grpc.Status;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The stats and tracing information for a call.
+ */
+public final class StatsTraceContext {
+  public static final StatsTraceContext NOOP = StatsTraceContext.newClientContext(
+      "noopservice/noopmethod", NoopCensusContextFactory.INSTANCE,
+      GrpcUtil.STOPWATCH_SUPPLIER);
+
+  private enum Side {
+    CLIENT, SERVER
+  }
+
+  private final CensusContext censusContext;
+  private final Stopwatch stopwatch;
+  private final Side side;
+  private final Metadata.Key<CensusContext> censusHeader;
+  private long wireBytesSent;
+  private long wireBytesReceived;
+  private long uncompressedBytesSent;
+  private long uncompressedBytesReceived;
+  private boolean callEnded;
+
+  private StatsTraceContext(Side side, String methodName, CensusContext parentContext,
+      Supplier<Stopwatch> stopwatchSupplier, Metadata.Key<CensusContext> censusHeader) {
+    this.side = side;
+    TagKey methodTagKey =
+        side == Side.CLIENT ? RpcConstants.RPC_CLIENT_METHOD : RpcConstants.RPC_SERVER_METHOD;
+    this.censusContext = parentContext.with(methodTagKey, new TagValue(methodName));
```

Comment:
Please add a TODO for me to cache this in the method descriptor.  This is going to create garbage and be slow.